---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# polyafit

<!-- badges: start -->
<!-- badges: end -->


`polyafit` is an R package that provides tools to fit count data to a
[Dirichlet-multinomial distribution](https://en.wikipedia.org/wiki/Dirichlet-multinomial_distribution).
This distribution is sometimes called the multivariate PÃ³lya distribution, from
which the package name is derived. The package was created specifically for
16S rRNA marker gene sequence data from the human microbiome, but in principle
could be applied to other types of data.

## Installation

You can install polyafit from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("kylebittinger/polyafit")
```

## Example

```{r, echo=FALSE, message=FALSE}
devtools::load_all()
```

```{r message=FALSE}
library(tidyverse)
```

We will simulate two samples, and check to see if any features are
different, based on the distribution of all the features in the samples.  We
assume that most of the features come from the same underlying distribution, so
we can use the full set of features to estimate the amount that feature
proportions vary between samples.

First, we'll generate data for both samples using the same underlying
distribution.  We will randomly generate some proportions using the 
Chi-Squared distribution.  The numbers won't add to 1, so we'll use a
convenience function to divide by the sum.

```{r}
as_proportion <- function (x) x / sum(x)
set.seed(1)
example_props <- as_proportion(rchisq(50, 2))
```

Now, we'll use a multinomial distribution to draw two samples with the same
underlying proportions.  We'll call these samples lung and OW, to represent
lung (lung) and oral (oral) microbiome samples from the same individual.

Just to demonstrate that our method doesn't require the same number of counts
in each sample, we'll draw a different number of items for each sample.

```{r}
lung_cts <- rmultinom(1, 500, example_props)
oral_cts <- rmultinom(1, 600, example_props)
```

Finally, we'll remove features with fewer than 5 counts total.  This method does
not work well for features that appear sparsely in the sample set.

```{r}
keep <- (lung_cts + oral_cts) >= 5
lung_cts <- lung_cts[keep]
oral_cts <- oral_cts[keep]
```

### Analysis: no outlier

Run an outlier analysis on the counts.  Here, the features all come from the
same underlying distribution, and we expect to see no outliers.

```{r}
tibble(Lung = lung_cts, Oral = oral_cts) %>%
  mutate(across(everything(), as_proportion)) %>%
  ggplot(aes(y=Lung, x=Oral)) +
  geom_point() +
  geom_abline(slope=1, intercept=0, linetype="dashed") +
  scale_x_log10() +
  scale_y_log10() +
  coord_equal()
```

Raw p-values for each feature.  All p-values are > 0.05.

```{r}
pfit <- optim_polya(rbind(lung_cts, oral_cts))
pvals <- 1 - ppolya_marginal(lung_cts, pfit$par, log.p = FALSE)
pvals
```

The fitted object, `pfit`, has a vector of estimated parameter values for the
Dirichlet-multinomial (DM) distribution, which serves as the mathematical model
for our data. From the estimated parameter values, we can compute an
overdispersion parameter, theta. Theta tells us how much "extra variance" the
estimated distribution has, beyond what we expect from a multinomial. Because
the data were generated from a perfect multinomial distribution, the estimated
overdispersion parameter should be very small. 

```{r}
1 / sum(pfit$par)
```

### Analysis: one outlier

Adjust one of the counts to be different in lung. Put in 150 counts for the 4th
feature.

```{r}
lung_cts[4] <- 150
```

```{r}
tibble(Lung = lung_cts, Oral = oral_cts) %>%
  mutate(across(everything(), as_proportion)) %>%
  ggplot(aes(y=Lung, x=Oral)) +
  geom_point() +
  geom_abline(slope=1, intercept=0, linetype="dashed") +
  scale_x_log10() +
  scale_y_log10() +
  coord_equal()
```

Doing the same analysis, we can see that the p-value for the 4th feature is
very very low (<< 0.05).

```{r}
pfit <- optim_polya(rbind(lung_cts, oral_cts))
pvals <- 1 - ppolya_marginal(lung_cts, pfit$par, log.p = FALSE)
pvals
```

Now, we re-calculate the overdispersion parameter, theta. It has increased by
three orders of magnitude.

```{r}
1 / sum(pfit$par)
```


